# EXERCISE

# 01. Importing libraries

In [10]:
import pandas as pd
import numpy as np
import os

# 02. Importing datasets

In [11]:
# define the path and assign dataframes to the "df" variables
path = r'C:\Users\Jakov\2023 Instacart Basket Analysis'
vars_list = ['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day', 'days_since_prior_order']
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False, usecols = vars_list)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [12]:
df_prods.shape

(49693, 5)

In [13]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [40]:

df_ords['order_id'] = df_ords['order_id'].astype('str')

In [6]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_a_week  order_hour_of_day  \
0   2539329        1             1                     2                  8   
1   2398795        1             2                     3                  7   
2    473747        1             3                     3                 12   
3   2254736        1             4                     4                  7   
4    431534        1             5                     4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [41]:
df_dep_t = df_dep.T # transpose the dataframe

In [42]:
df_dep_t.reset_index() # create the index column

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [43]:
new_header = df_dep_t.iloc[0] # store the row with index 0 in the variable
df_dep_t_new = df_dep_t[1:]
df_dep_t_new.columns = new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [44]:
data_dict = df_dep_t_new.to_dict('index') # turn departments dataframe into data dictionary

In [45]:
df_snacks = df_prods[df_prods['department_id'] == 19] # data subsetting

In [21]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# TASK

# 2.

In [46]:
df_ords['user_id'] = df_ords['user_id'].astype('str') # turn 'user_id' column into string object

# 3.

In [48]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_a_week'}, inplace = True) # rename the 'order_dow' column

# 4. Answer - 10:00 is the busiest hour for placing orders.

In [52]:
df_ords['order_hour_of_day'].value_counts(dropna = False) # show count of each value

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

# 5. Answer - Value 4 in 'department_id' column stands for 'produce'.

In [53]:
print(data_dict.get('4')) # retrieve an object with the key value of 4

{'department': 'produce'}


# 6.

In [57]:
data_dict # show data dictionary
df_breakfast = df_prods[df_prods['department_id'] == 14] # in the variable 'df_breakfast', store all rows with 'department_id' value of 14

In [56]:
df_breakfast # show 'df_breakfast' subset

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# 7.

In [58]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [59]:
df_alc_deli_bev_seaf = df_prods[df_prods['department_id'].isin([5, 7, 20, 12])] # create subset with rows with 'department_id' values of 5 , 7, 12, 20
df_alc_deli_bev_seaf # show created subset

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# 8. Answer - The last dataset that I created has 7650 rows.

In [63]:
df_alc_deli_bev_seaf.info() # show info about 'df_alc_deli_bev_seaf' dataframe

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7650 entries, 2 to 49688
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     7650 non-null   int64  
 1   product_name   7647 non-null   object 
 2   aisle_id       7650 non-null   int64  
 3   department_id  7650 non-null   int64  
 4   prices         7650 non-null   float64
dtypes: float64(1), int64(3), object(1)
memory usage: 358.6+ KB


# 9.

In [66]:
customer_id_1 = df_ords[df_ords['user_id'] == '1'] # create subset of 'df_ords' dataframe with rows with 'user_id' = 1
customer_id_1

order_id user_id  order_number  orders_day_of_a_week  order_hour_of_day  \
0   2539329       1             1                     2                  8   
1   2398795       1             2                     3                  7   
2    473747       1             3                     3                 12   
3   2254736       1             4                     4                  7   
4    431534       1             5                     4                 15   
5   3367565       1             6                     2                  7   
6    550135       1             7                     1                  9   
7   3108588       1             8                     1                 14   
8   2295261       1             9                     1                 16   
9   2550362       1            10                     4                  8   
10  1187899       1            11                     4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# 10. Answers - The user has 4 times placed an order on Thursdays, 3 times on Mondays and 2 times on Tuesdays and Wednesdays. The user has at 08:00 and 07:00 placed an order 3 times. At 09:00, 12:00, 14:00, 15:00 and 16:00, the user has placed only 1 order for each mentioned hour. Altogether user has placed 10 orders. On average, the user has had placed the orders every 19 days. The longest period without placing orders is 30 days, the shortest period is 0 days.

In [70]:
customer_id_1['orders_day_of_a_week'].value_counts(dropna = False) # show frequency table for the 'orders_day_of_a_week' column

4    4
1    3
2    2
3    2
Name: orders_day_of_a_week, dtype: int64

In [71]:
customer_id_1['order_hour_of_day'].value_counts(dropna = False) # show frequency table for the 'order_hour_of_Day' column

8     3
7     3
12    1
15    1
9     1
14    1
16    1
Name: order_hour_of_day, dtype: int64

In [73]:
customer_id_1.describe() # show statistical description of the 'customer_id_1' dataframe

order_number  orders_day_of_a_week  order_hour_of_day  \
count     11.000000             11.000000          11.000000   
mean       6.000000              2.636364          10.090909   
std        3.316625              1.286291           3.477198   
min        1.000000              1.000000           7.000000   
25%        3.500000              1.500000           7.500000   
50%        6.000000              3.000000           8.000000   
75%        8.500000              4.000000          13.000000   
max       11.000000              4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# 12. 

In [74]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

# 13.

In [76]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))

In [7]:
df_ords.shape

(3421083, 6)